### DISCUSSION TITLE
GDMP for GODOT 3.5- windows

### DISCUSSION THREAD

**Discussion by Ehsanwww** [j20001970/GDMP#26] [General]
Hello everyone, thanks for this great asset.
I want to compile mediapipe for use in godot 3.5 but I got Error bellow when compiling:
"
ERROR: D:/gdmp/mediapipe/GDMP/desktop/BUILD:1:10: Linking GDMP/desktop/GDMP.dll failed: (Exit 1120): link.exe failed: error executing command (from target //GDMP/desktop:GDMP) D:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.29.30133\bin\HostX64\x64\link.exe @bazel-out/x64_windows-opt/bin/GDMP/desktop/GDMP.dll-2.params
LINK : warning LNK4044: unrecognized option '/s'; ignored
"

**Comment by j20001970** [comment]
> ERROR: D:/gdmp/mediapipe/GDMP/desktop/BUILD:1:10: Linking GDMP/desktop/GDMP.dll failed: (Exit 1120): link.exe failed: ...

Please provide relevant log regarding linking error. The '/s' option that are used in Linux/Android is ignored when building for Windows library, and is not the cause of linkage failure.

**Comment by Ehsanwww** [comment]
@j20001970 
ERROR: D:/gdmp/mediapipe/GDMP/desktop/BUILD:1:10: Linking GDMP/desktop/GDMP.dll failed: (Exit 1120): link.exe failed: error executing command (from target //GDMP/desktop:GDMP) D:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.29.30133\bin\HostX64\x64\link.exe @bazel-out/x64_windows-opt/bin/GDMP/desktop/GDMP.dll-2.params
LINK : warning LNK4044: unrecognized option '/s'; ignored
   Creating library bazel-out/x64_windows-opt/bin/GDMP/desktop/GDMP.if.lib and object bazel-out/x64_windows-opt/bin/GDMP/desktop/GDMP.if.exp
LINK : warning LNK4217: symbol 'TfLiteXNNPackDelegateOptionsDefault' defined in 'xnnpack_delegate.lib(xnnpack_delegate.obj)' is imported by 'inference_calculator_cpu.lo.lib(inference_calculator_cpu.obj)' in function '"private: class absl::lts_20230125::StatusOr<class std::unique_ptr<struct TfLiteDelegate,class std::function<void __cdecl(struct TfLiteDelegate *)> > > __cdecl mediapipe::api2::InferenceCalculatorCpuImpl::MaybeCreateDelegate(class mediapipe::CalculatorContext *)" (?MaybeCreateDelegate@InferenceCalculatorCpuImpl@api2@mediapipe@@AEAA?AV?$StatusOr@V?$unique_ptr@UTfLiteDelegate@@V?$function@$$A6AXPEAUTfLiteDelegate@@@Z@std@@@std@@@lts_20230125@absl@@PEAVCalculatorContext@3@@Z)'
LINK : warning LNK4217: symbol 'TfLiteXNNPackDelegateCreate' defined in 'xnnpack_delegate.lib(xnnpack_delegate.obj)' is imported by 'inference_calculator_cpu.lo.lib(inference_calculator_cpu.obj)' in function '"private: class absl::lts_20230125::StatusOr<class std::unique_ptr<struct TfLiteDelegate,class std::function<void __cdecl(struct TfLiteDelegate *)> > > __cdecl mediapipe::api2::InferenceCalculatorCpuImpl::MaybeCreateDelegate(class mediapipe::CalculatorContext *)" (?MaybeCreateDelegate@InferenceCalculatorCpuImpl@api2@mediapipe@@AEAA?AV?$StatusOr@V?$unique_ptr@UTfLiteDelegate@@V?$function@$$A6AXPEAUTfLiteDelegate@@@Z@std@@@std@@@lts_20230125@absl@@PEAVCalculatorContext@3@@Z)'
LINK : warning LNK4217: symbol 'TfLiteXNNPackDelegateDelete' defined in 'xnnpack_delegate.lib(xnnpack_delegate.obj)' is imported by 'inference_calculator_cpu.lo.lib(inference_calculator_cpu.obj)' in function '"private: class absl::lts_20230125::StatusOr<class std::unique_ptr<struct TfLiteDelegate,class std::function<void __cdecl(struct TfLiteDelegate *)> > > __cdecl mediapipe::api2::InferenceCalculatorCpuImpl::MaybeCreateDelegate(class mediapipe::CalculatorContext *)" (?MaybeCreateDelegate@InferenceCalculatorCpuImpl@api2@mediapipe@@AEAA?AV?$StatusOr@V?$unique_ptr@UTfLiteDelegate@@V?$function@$$A6AXPEAUTfLiteDelegate@@@Z@std@@@std@@@lts_20230125@absl@@PEAVCalculatorContext@3@@Z)'
gesture_recognizer.lo.lib(gesture_recognizer.obj) : error LNK2019: unresolved external symbol "public: void __cdecl MediaPipeGestureRecognizerResult::_init(void)" (?_init@MediaPipeGestureRecognizerResult@@QEAAXXZ) referenced in function "void * __cdecl godot::_godot_class_instance_func<class MediaPipeGestureRecognizerResult>(void *,void *)" (??$_godot_class_instance_func@VMediaPipeGestureRecognizerResult@@@godot@@YAPEAXPEAX0@Z)
bazel-out\x64_windows-opt\bin\GDMP\desktop\GDMP.dll : fatal error LNK1120: 1 unresolved externals
Target //GDMP/desktop:GDMP failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 675.725s, Critical Path: 452.89s
INFO: 2913 processes: 7 internal, 2906 local.
FAILED: Build did NOT complete successfully

**Reply by j20001970** [reply]
This error is probably caused by missing MediaPipeGestureRecognizer::_init symbol that is fixed by recent rebase, please pull latest 3.x branch ([ee6d227](https://github.com/j20001970/GDMP/commit/ee6d22706dd6109cdebd8b3a7b5e55e7c72c031b)) then try compiling again.

**Comment by Ehsanwww** [comment]
@j20001970 , First i get this error:

cl : Command line warning D9002 : ignoring unknown option '-O3'
ERROR: D:/gdmp/mediapipe/GDMP/gpu/BUILD:1:11: Compiling GDMP/gpu/gpu_helper.cc failed: (Exit 2): cl.exe failed: error executing command (from target //GDMP/gpu:gpu_helper) D:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.29.30133\bin\HostX64\x64\cl.exe /nologo /DCOMPILER_MSVC /DNOMINMAX /D_WIN32_WINNT=0x0601 /D_CRT_SECURE_NO_DEPRECATE ... (remaining 54 arguments skipped)
GDMP/gpu/gpu_helper.cc(10): error C2039: 'get_gpu_buffer': is not a member of 'MediaPipeGPUHelper'
C:\users\ehsanak\_bazel_ehsanak\c3on7xoo\execroot\mediapipe\GDMP\gpu\gpu_helper.h(19): note: see declaration of 'MediaPipeGPUHelper'
GDMP/gpu/gpu_helper.cc(15): error C2065: 'gpu_helper': undeclared identifier
GDMP/gpu/gpu_helper.cc(16): error C2065: 'gpu_helper': undeclared identifier
GDMP/gpu/gpu_helper.cc(16): error C3482: 'this' can only be used as a lambda capture within a non-static member function
Target //GDMP/desktop:GDMP failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 412.255s, Critical Path: 350.35s
INFO: 1351 processes: 133 internal, 1218 local.

==================
then I comment gpu_helper.cc like what you have done in GDMP 4, but I got first error.

**Reply by Ehsanwww** [reply]
I got these two warnings when building for desktop 

WARNING: Download from https://storage.googleapis.com/mirror.tensorflow.org/github.com/pytorch/cpuinfo/archive/87d8234510367db49a65535021af5e1838a65ac2.zip failed: class java.io.FileNotFoundException GET returned 404 Not Found
WARNING: D:/project_git/gdmp3/mediapipe/mediapipe/framework/tool/BUILD:200:24: in cc_library rule //mediapipe/framework/tool:field_data_cc_proto: target '//mediapipe/framework/tool:field_data_cc_proto' depends on deprecated target '@com_google_protobuf//:cc_wkt_protos': Only for backward compatibility. Do not use.
WARNING: D:/project_git/gdmp3/mediapipe/mediapipe/framework/BUILD:59:24: in cc_library rule //mediapipe/framework:calculator_cc_proto: target '//mediapipe/framework:calculator_cc_proto' depends on deprecated target '@com_google_protobuf//:cc_wkt_protos': Only for backward compatibility. Do not use.

**Reply by j20001970** [reply]
Forgot to exclude MediaPipeGPUHelper::get_gpu_buffer from CPU-only build, fixed by [fcedd7d](https://github.com/j20001970/GDMP/commit/fcedd7d81480639537efac747fdc02e306cb8188).

**Reply by Ehsanwww** [reply]
@j20001970 
after fetching and doing above comments, I got this error:
ERROR: D:/project_git/gdmp3/mediapipe/GDMP/gpu/BUILD:1:11: Compiling GDMP/gpu/gpu_helper.cc failed: (Exit 2): cl.exe failed: error executing command (from target //GDMP/gpu:gpu_helper) D:\Program Files (x86)\Microsoft Visual Studio\2019\BuildTools\VC\Tools\MSVC\14.29.30133\bin\HostX64\x64\cl.exe /nologo /DCOMPILER_MSVC /DNOMINMAX /D_WIN32_WINNT=0x0601 /D_CRT_SECURE_NO_DEPRECATE ... (remaining 54 arguments skipped)
.\mediapipe/objc/util.h(18): fatal error C1083: Cannot open type library file: 'Accelerate/Accelerate.h': No such file or directory

**Reply by j20001970** [reply]
The compiler includes `mediapipe/objc/util.h` that should only be used when building for Apple platforms, have no idea how that happen, but can you try moving line 6-8 of gpu_helper.cc into `#if !MEDIAPIPE_DISABLE_GPU` macro?

**Reply by Ehsanwww** [reply]
![gpu](https://github.com/j20001970/GDMP/assets/20922281/4bbe1eeb-05ba-4fbe-a050-bec55f410109)

**Reply by Ehsanwww** [reply]
@j20001970 be default, it is inside that macro

**Reply by Ehsanwww** [reply]
but it gives that error @j20001970

**Reply by j20001970** [reply]
Could you update the information about which error occur? Does `'get_gpu_buffer': is not a member of 'MediaPipeGPUHelper'` still exist, or it is `Cannot open type library file: 'Accelerate/Accelerate.h': No such file or directory` that fails the compilation?
